Lib/genericpath.py

```diff
@@ -45,9 +45,20 @@ def isdir(s):
     return stat.S_ISDIR(st.st_mode)
 
 
-def getsize(filename):
-    """Return the size of a file, reported by os.stat()."""
-    return os.stat(filename).st_size
+def getsize(filename, apparent=True):
+    """Return the size of a file, reported by os.stat().
+
+    If 'apparent' is True (default), the apparent size (number of bytes) of the
+    file is returned. If False, the actual size (disk space occupied) is
+    returned. The actual size reflects the block size, meaning it will
+    typically be larger than the apparent size. However, the inverse may also
+    be true due to holes in ("sparse") files, internal fragmentation, indirect
+    blocks, etc.
+    """
+    if apparent:
+        return os.stat(filename).st_size
+    else:
+        return os.stat(filename).st_blocks * 512
```

Comment:
Where is this magic constant coming from? Are you assuming that all file systems and operating systems have 512 bytes block size? POSIX standard does not specify the block size, see https://pubs.opengroup.org/onlinepubs/009695399/basedefs/sys/stat.h.html

Reply:
I'm taking it from [`fstat(2)`](https://linux.die.net/man/2/fstat):

> The st_blocks field indicates the number of blocks allocated to the file, 512-byte units. (This may be smaller than st_size/512 when the file has holes.)

[Our own docs](https://docs.python.org/3/library/os.html#os.stat_result.st_blocks) state much the same thing.

> **st_blocks**
> Number of 512-byte blocks allocated for file. This may be smaller than st_size/512 when the file has holes.

Note that we should not use `st_blksize` because it's the preferred blocksize, not the actual blocksize. From `fstat(2)`:

> The st_blksize field gives the "preferred" blocksize for efficient file system I/O. (Writing to a file in smaller chunks may cause an inefficient read-modify-rewrite.)

Would a (private) constant be beneficial? Or simply a comment, perhaps?

Reply:
The man page https://linux.die.net/man/2/fstat reflects the situation on Linux. Python supports many more platforms than just Linux or BSD. I guess our own documentation is also based on the Linux or BSD documentation, too.

The Open Group docs cover the generic POSIX standard. It mentions that some platforms define ``DEV_BSIZE`` in ``<sys/param.h>``. I also found some code that defines ``DEV_BSIZE`` from ``BSIZE``. I recommend to expose the constant from ``posixmodule.c`` and only support the feature when the system has a usable ``DEV_BSIZE`` or ``BSIZE``.